patches/api/0421-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,2335 @@
++     *
++     * @param fireball The fireball
++     * @return         The damage origin created
++     * @see VanillaDamageType#FIREBALL
++     */
++    @NotNull
++    public static DamageOrigin ofFireball(@NotNull Fireball fireball) {
++        return ofFireball(fireball, null);
++    }
++
++    /**
++     * Fireball damage origin is dealt when a fireball explode on a target.
++     * Wither skulls should use the {@link #ofWitherSkull(WitherSkull, Entity)} instead.
++     *
++     * @param fireball The projectile
++     * @param thrower  The thrower
++     * @return         The damage origin created
++     * @see VanillaDamageType#FIREBALL
++     */
++    @NotNull
++    public static DamageOrigin ofFireball(@NotNull Fireball fireball, @Nullable Entity thrower) {
++        if (fireball instanceof WitherSkull) { // Thanks bukkit
++            throw new IllegalArgumentException("Projectile can only be a variant of fireball. Wither skull has its own method.");
++        }
++
++        return thrower == null ? ofType(VanillaDamageType.UNATTRIBUTED_FIREBALL, fireball) : ofType(VanillaDamageType.FIREBALL, fireball, thrower);
++    }
++
++    /**
++     * Wither skull damage origin is dealt when a wither skull explode on a target.
++     *
++     * @param witherSkull The wither skull
++     * @return            The damage origin created
++     * @see VanillaDamageType#WITHER_SKULL
++     */
++    @NotNull
++    public static DamageOrigin ofWitherSkull(@NotNull WitherSkull witherSkull) {
++        return ofWitherSkull(witherSkull, null);
++    }
++
++    /**
++     * Wither skull damage origin is dealt when a wither skull explode on a target.
++     *
++     * @param witherSkull The wither skull
++     * @param thrower     The thrower
++     * @return            The damage origin created
++     * @see VanillaDamageType#WITHER_SKULL
++     */
++    @NotNull
++    public static DamageOrigin ofWitherSkull(@NotNull WitherSkull witherSkull, @Nullable Entity thrower) {
++        return ofType(VanillaDamageType.WITHER_SKULL, witherSkull, thrower);
++    }
++
++    /**
++     * Passive projectile damage origin is dealt when a projectile hits a target
++     * but will affect only sensitive mobs.
++     *
++     * @param projectile The projectile
++     * @return           The damage origin created
++     * @see VanillaDamageType#THROWN
++     */
++    @NotNull
++    public static DamageOrigin ofPassiveProjectile(@NotNull Entity projectile) {
++        return ofPassiveProjectile(projectile, null);
++    }
++
++    /**
++     * Passive projectile damage origin is dealt when a projectile hits a target
++     * but will affect only sensitive mobs.
++     *
++     * @param projectile The projectile
++     * @param thrower    The thrower
++     * @return           The damage origin created
++     * @see VanillaDamageType#THROWN
++     */
++    @NotNull
++    public static DamageOrigin ofPassiveProjectile(@NotNull Entity projectile, @Nullable Entity thrower) {
```

Comment:
Projectile?

Reply:
This was the old name of that method and it make sense if we see that it's only used for passive projectile like egg but i can just use ofThrown if you prefer

Reply:
That may be a better name, but I was more referring to the param ``Entity projectile``